### DISCUSSION TITLE
Issue with Lets Encrypt

### DISCUSSION THREAD

**Discussion by jayb323** [requarks/wiki#3040] [Help / Questions]
Hi all,

I've installed wiki.js on Ubuntu 20.04, PostgreSQL, and NGINX. 

I had it working up until the point where I tried to change the port so users didn't need to add ":3000" at the end, but now, since changing it (I tried 443 to begin with, then 3443) it gives me the following SSL error.

```Loading configuration from /var/www/wiki.example.com/html/config.yml... OK
2021-02-07T05:51:46.464Z [MASTER] info: =======================================
2021-02-07T05:51:46.466Z [MASTER] info: = Wiki.js 2.5.170 =====================
2021-02-07T05:51:46.466Z [MASTER] info: =======================================
2021-02-07T05:51:46.467Z [MASTER] info: Initializing...
2021-02-07T05:51:47.082Z [MASTER] info: Using database driver pg for postgres [ OK ]
2021-02-07T05:51:47.088Z [MASTER] info: Connecting to database...
2021-02-07T05:51:47.114Z [MASTER] info: Database Connection Successful [ OK ]
2021-02-07T05:51:47.768Z [MASTER] warn: Mail is not setup! Please set the configuration in the administration area!
2021-02-07T05:51:47.876Z [MASTER] info: Loading GraphQL Schema...
2021-02-07T05:51:48.798Z [MASTER] info: GraphQL Schema: [ OK ]
2021-02-07T05:51:49.109Z [MASTER] info: HTTP Server on port: [ 3443 ]
2021-02-07T05:51:49.143Z [MASTER] info: (LETSENCRYPT) Initializing Let's Encrypt client...
2021-02-07T05:51:49.176Z [MASTER] info: HTTP Server: [ RUNNING ]
2021-02-07T05:51:49.947Z [MASTER] info: (LETSENCRYPT) Generating server keypairs...
2021-02-07T05:51:50.080Z [MASTER] info: (LETSENCRYPT) Server keypairs generated successfully [ OK ]
2021-02-07T05:51:50.081Z [MASTER] info: (LETSENCRYPT) Generating certificate signing request (CSR)...
2021-02-07T05:51:50.087Z [MASTER] info: (LETSENCRYPT) CSR generated successfully [ OK ]
2021-02-07T05:51:50.087Z [MASTER] info: (LETSENCRYPT) Requesting certificate from Let's Encrypt...
2021-02-07T05:51:50.091Z [MASTER] info: (LETSENCRYPT) Setting HTTP challenge for wiki.example.com: [ READY ]
2021-02-07T05:51:50.092Z [MASTER] info: (LETSENCRYPT) Waiting for challenge to complete...
2021-02-07T05:51:55.106Z [MASTER] info: (LETSENCRYPT) Removing HTTP challenge: [ OK ]
2021-02-07T05:51:55.106Z [MASTER] info: (LETSENCRYPT) Removing HTTP challenge: [ OK ]
2021-02-07T05:51:55.106Z [MASTER] warn: (LETSENCRYPT) Error: Error: Failed HTTP-01 Pre-Flight / Dry Run.
curl 'http://wiki.example.com/.well-known/acme-challenge/test-b00f6923aaa23017e15401e5a13b5ff3-0'
Expected: 'test-b00f6923aaa23017e15401e5a13b5ff3-0.lFZVzcLbOKAHMVmrk5d1wr0Rsvqwpx3SzECqhkDApqg'
Got: '<html>
<head><title>404 Not Found</title></head>
<body>
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.18.0 (Ubuntu)</center>
</body>
</html>
'
See https://git.rootprojects.org/root/acme.js/issues/4
2021-02-07T05:51:55.108Z [MASTER] error: Error: Failed HTTP-01 Pre-Flight / Dry Run.
curl 'http://wiki.example.com/.well-known/acme-challenge/test-b00f6923aaa23017e15401e5a13b5ff3-0'
Expected: 'test-b00f6923aaa23017e15401e5a13b5ff3-0.lFZVzcLbOKAHMVmrk5d1wr0Rsvqwpx3SzECqhkDApqg'
Got: '<html>
<head><title>404 Not Found</title></head>
<body>
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.18.0 (Ubuntu)</center>
</body>
</html>
'
See https://git.rootprojects.org/root/acme.js/issues/4
```

This is my config.yml

```#######################################################################
# Wiki.js - CONFIGURATION                                             #
#######################################################################
# Full documentation + examples:
# https://docs.requarks.io/install

# ---------------------------------------------------------------------
# Port the server should listen to
# ---------------------------------------------------------------------

port: 3443

# ---------------------------------------------------------------------
# Database
# ---------------------------------------------------------------------
# Supported Database Engines:
# - postgres = PostgreSQL 9.5 or later
# - mysql = MySQL 8.0 or later (5.7.8 partially supported, refer to docs)
# - mariadb = MariaDB 10.2.7 or later
# - mssql = MS SQL Server 2012 or later
# - sqlite = SQLite 3.9 or later

db:
  type: postgres

  # PostgreSQL / MySQL / MariaDB / MS SQL Server only:
  host: localhost
  port: 5432
  user: wikijs
  pass: ***hidden***
  db: wikijs
  ssl: false

  # Optional - PostgreSQL / MySQL / MariaDB only:
  # -> Uncomment lines you need below and set `auto` to false
  # -> Full list of accepted options: https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_optio>
  sslOptions:
    auto: true
    # rejectUnauthorized: false
    # ca: path/to/ca.crt
    # cert: path/to/cert.crt
    # key: path/to/key.pem
    # pfx: path/to/cert.pfx
    # passphrase: xyz123

  # SQLite only:
  storage: path/to/database.sqlite

#######################################################################
# ADVANCED OPTIONS                                                    #
#######################################################################
# Do not change unless you know what you are doing!

# ---------------------------------------------------------------------
# SSL/TLS Settings
# ---------------------------------------------------------------------
# Consider using a reverse proxy (e.g. nginx) if you require more
# advanced options than those provided below.

ssl:
  enabled: true
  port: 3443

  # Provider to use, possible values: custom, letsencrypt
  provider: letsencrypt

  # ++++++ For custom only ++++++
  # Certificate format, either 'pem' or 'pfx':
  format: pem
  # Using PEM format:
  key: path/to/key.pem
  cert: path/to/cert.pem
  # Using PFX format:
  pfx: path/to/cert.pfx
  # Passphrase when using encrypted PEM / PFX keys (default: null):
  passphrase: null
  # Diffie Hellman parameters, with key length being greater or equal
  # to 1024 bits (default: null):
  dhparam: null

  # ++++++ For letsencrypt only ++++++
  domain: wiki.example.com
  subscriberEmail: email@gmail.com

# ---------------------------------------------------------------------
# Database Pool Options
# ---------------------------------------------------------------------
# Refer to https://github.com/vincit/tarn.js for all possible options

pool:
  # min: 2
  # max: 10

# ---------------------------------------------------------------------
# IP address the server should listen to
# ---------------------------------------------------------------------
# Leave 0.0.0.0 for all interfaces

bindIP: 0.0.0.0

# ---------------------------------------------------------------------
# Log Level
# ---------------------------------------------------------------------
# Possible values: error, warn, info (default), verbose, debug, silly

logLevel: info

# ---------------------------------------------------------------------
# Offline Mode
# ---------------------------------------------------------------------
# If your server cannot access the internet. Set to true and manually
# download the offline files for sideloading.

offline: false

# ---------------------------------------------------------------------
# High-Availability
# ---------------------------------------------------------------------
# Set to true if you have multiple concurrent instances running off the
# same DB (e.g. Kubernetes pods / load balanced instances). Leave false
# otherwise. You MUST be using PostgreSQL to use this feature.

ha: false

# ---------------------------------------------------------------------
# Data Path
# ---------------------------------------------------------------------
# Writeable data path used for cache and temporary user uploads.
dataPath: ./data
```

NGINX config for example.com

```server {

        root /var/www/wiki.example.com/html;
        index index.html index.htm index.nginx-debian.html;

        server_name wiki.example.com;

        location / {
                try_files $uri $uri/ =404;
        }


    listen [::]:443 ssl ipv6only=on; # managed by Certbot
    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/wiki.example.com/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/wiki.example.com/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot


}
server {
    if ($host = wiki.example.com) {
        return 301 https://$host$request_uri;
    } # managed by Certbot


        listen 80;
        listen [::]:80;

        server_name wiki.example.com www.wiki.example.com;
    return 404; # managed by Certbot


}
```

Any ideas?

**Comment by jayb323** [comment]
Anyone, or would I be best of posting on an Ubuntu forum?

**Reply by justinacolmena** [reply]
I have been using a similiar setup -- I already posted about it here. https://github.com/Requarks/wiki/issues/2155#issuecomment-655253373

I just up a new server with nginx & certbot, quick and easy for automated setup, but there are some odd settings that may need manual adjustment.

    listen [::]:443 ssl ipv6only=on; # managed by Certbot
    listen 443 ssl; # managed by Certbot

I've had trouble with certbot. The cron job sometimes fails to renew the certs, and certain settings are being clobbered or "overmanaged."  Using `certbot certonly` isn't supposed to clobber settings, I believe.

**Reply by mattpr** [reply]
Yes.  For full control I use certonly and then as long as certbot succeeds I know where to find my key/cert to point nginx config at the files.  So my nginx config can be managed by me and not auto-modified by certbot or anyone else.

We also use our own system to check certs for upcoming expiration and once within a certain window of expiration we trigger the renew.  Renew is no different than just running a new certbot `certonly` command.  The `renew` in certbot is a convenience to retain previous used config (domain names, etc) so you don't have to re-enter them.  But if you are maintaining all this config already outside of certbot, then there is no reason to use the certbot renew over just creating a fresh new certificate.

If you are using a monitoring system like prometheus with blackbox exporter it is pretty easy to get alerted when certificates on your http endpoints are within X days of expiration (ie certbot is failing for some reason).

**Comment by mattpr** [comment]
```
curl 'http://wiki.example.com/.well-known/acme-challenge/test-b00f6923aaa23017e15401e5a13b5ff3-0'
Expected: 'test-b00f6923aaa23017e15401e5a13b5ff3-0.lFZVzcLbOKAHMVmrk5d1wr0Rsvqwpx3SzECqhkDApqg'
Got: '<html>
<head><title>404 Not Found</title></head>
```
Let's encrypt gives you a free DV (domain verification) cert only after they verify you control the domain.  They are making a request to a file they have generated which they expect to be here: `http://wiki.example.com/.well-known/acme-challenge/test-b00f6923aaa23017e15401e5a13b5ff3-0` but instead of getting the file that verifies you control the site/domain, they get a 404 (not found) error so verification fails.  So the file isn't there or nginx is looking in the wrong place or if wiki.js is supposed to serve the file instead of nginx, the wiki.js is looking in the wrong place.

Did you change the domain to wiki.example.com to hide your own?  If not, that is definitely the problem.

LOTS of reasons this can happen.  Here are a few off the top of my head.

- letsencrypt/certbot is checking wrong domain (e.g. wiki.example.com when your site is at wiki.mydomain.com)
- letsencrypt/certbot (or wiki.js that configures certbot) is pointing at wrong web_root so this file `/.well-known/acme-challenge/test-b00f6923aaa23017e15401e5a13b5ff3-0` is created in wrong place (should be created in `WEBROOT/.well-known/acme-challenge/test-b00f6923aaa23017e15401e5a13b5ff3-0`).
- nginx is pointing at wrong webroot or not configured to serve files and only does upstream proxy to wiki.js (`root` directive in nginx server block)
- wiki.js is supposed to serve file instead of nginx but is looking for the file (`/.well-known/acme-challenge/test-b00f6923aaa23017e15401e5a13b5ff3-0`) in the wrong place

If you run certbot manually you can put it into a debug mode to stop mid verification so you can check and verify the verification tokens are where you expect them to be on the filesystem.  Once you verify that you need to check that it isn't nginx's fault.  Manually load up wiki.mydomain.com/.well-known... from certbot output in your own browser or with curl.  Do you get the verification token or a 404 error?  If you get the token then the generation/nginx/serving is fine and it is most likely that certbot/letsencrypt is looking in the wrong place (bad domain?  told wiki.js wrong domain?)

Ubuntu forum won't help.
Problem is certainly somewhere between letsencrypt/certbot/nginx.  Wiki.js's only role here is the config it is passing to run certbot for letsencrypt which could be wrong.  

If you already have certificates that are valid for your domain, you can just turn off the letsencrypt/certbot stuff in wiki.js config, configure ssl in your nginx conf to use those certs and just use wiki.js as a non-ssl upstream.  And maybe figure out how to run certbot yourself on ubuntu to update certs as needed (they are only valid for 90 days but there is an autorenew function built into certbot...worth learning).

**Reply by jayb323** [reply]
@mattpr Feel like consulting/doing a paid job?

**Reply by mattpr** [reply]
Hi Jay, I appreciate the sentiment but have more work than I can stay ahead of already.
I hammered out the above answer in a hurry so apologies for grammar and inconsistency in terms (letsencrypt is CA/service, certbot is the popular "get a cert" client app that runs on your server and talks to letsencrypt).

The answer is there.  This isn't really too hard if you are already able to setup nginx, install wiki.js, etc.  

**You just need to understand how the components are interacting or depending on common paths on disk.**
- certbot is putting files on disk somewhere (where?  running certbot in debug mode can help.).  
- Nginx is serving files for some domain from some directory (or not)...make sure it is serving files (in particular from `WEB_ROOT/.well-known/...` and you know what directory it thinks `WEB_ROOT` is.  
- Can you save a file in that directory on disk and then access it in the web browser over http using the expected domain and path to the file under that directory?  
    - If not...get that working first.  
    - Once that is working then it is just a matter of making sure where letsencrypt is looking (protocol/domain/path) matches up with what nginx is serving and the files on disk.
    - You can run certbot in debug mode to stop after the verification files are generated, before doing verification and before deleting those files.  This allows you to verify they exist on disk where you think and also to test whether you can load them yourselves using a web browser at the expected URL...if not, letsencrypt can't either.

I do SSL termination on nginx end anyway and am using certbot directly through our server provisioning code.  Wiki.js is just an http upstream that doesn't need to mess with any of that (in my case).   

Another thing...letsencrypt will follow redirects for verification.  So if they try: `http://www.example.com/.well-known/acme-challenge/VERIFICATION_KEY` and that redirects to `https://` they will follow that.  But that depends on already working https (valid cert already in place for nginx to use).  In my case to bootstrap servers, I actually have a temporary (pure http, no redirects or upstream) nginx config that is used before certbot runs which is replaced with the final http/https nginx config after certbot has successfully run.  Future cert renewals will work with the http/https config because a valid certificate is in place (and also because I have a special case location block on the http server for letsencrypt's benefit).

You can check what redirects your server is doing by doing `curl -L -i "URL"` to see the redirect headers and body.  Or open browser dev tools, open network panel, click the checkbox to "preserve log" (across page views), clear the current log then paste your url in the address bar and hit enter.  You can see redirects based on HTTP status (30_) and the presence of a `Location: ` http header on the response from the server.

One thing that might help keep things simple... very common to have http redirecting to https but if https is broken verification can't happen.  You might want to use an nginx config something like the following that will serve these verification tokens over http via a dedicated location block.  Again this only makes sense if you are terminating ssl on nginx rather than wiki.js.

But this only makes sense if you are terminating or plan to terminate https with nginx instead of wiki.js.

Good luck!

```
server {
    listen 80;
    listen [::]:80;

    server_name {{.DOMAIN}};

    root {{.WEB_ROOT}};

    # ensure letsencrypt can verify this domain when https is broken
    location /.well-known/acme-challenge/ {
        try_files $uri @redirect; 
    }

    # redirect to https
    location / {
        try_files /dev/null @redirect;
    }

    # redirect to https
    location @redirect {
        return 301 https://{{.DOMAIN}}$request_uri; 
    }
}
server {
    listen 443 ssl;
    listen [::]:443 ssl;

    ssl_certificate     certs/{{.TLS_CRT}};
    ssl_certificate_key certs/{{.TLS_KEY}};

    server_name {{.DOMAIN}};

    root {{.WEB_ROOT}};

    # more server conf...(gzip, static asset caching, tls support, etc)

    # make sure we are still willing to serve files for letsencrypt.  
    # only required if you aren't serving .well-known files in the http server block above.
    # This block says: try to find file at WEB_ROOT/$uri on disk.  If it exists, serve it, otherwise give the request to upstream (wiki.js)
    location {{.UPSTREAM_PATH_PREFIX}} {   # / for wiki.js
        try_files $uri @upstream;
    }

    location @upstream {
        proxy_pass http://localhost:{{.UPSTREAM_PORT}};
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        add_header  X-Upstream  js.wiki;
    }

}
```
